File: packages/butler-agent/src/agent/btcc/turn/runtime.ts
/** Durable Turn runtime implementation; the public Turn facade remains in turn.ts. */
import type {
  BtccRunCommand,
  BtccStopCommand,
  BtccTurnOutcome,
  BtccTurnProgressObserver,
  BtccTurnRuntime,
} from "../contracts.ts";
import type { CanonicalMessageStore } from "../delivery/index.ts";
import { insertCanonicalMessage } from "../delivery/index.ts";
import { createTurnExecutionSupervisor } from "../recovery/index.ts";
import type { CommittedSuccessorReadiness } from "../recovery/index.ts";
import { acquireStateExecution } from "./acquire-state-execution.ts";
import { loadOrAdmitTurn } from "./load-or-admit-turn.ts";
import { projectTerminalOutcome } from "./project-terminal-outcome.ts";
import type {
  TurnAdmissionRepository,
  TurnRecord,
  TurnStateRepository,
} from "./contracts.ts";
import { stopTurn } from "./stop-turn.ts";
import { isSqliteContention } from "../../../foundation/sqlite-contention.ts";
import type { BtccAgentLoop, BtccAgentLoopResult } from "../agent-loop/index.ts";
import { isGuidedWorkCloseoutError } from "../agent-loop/index.ts";
import {
  isPhaseContinuityProjectionError,
  isRoundToolSurfaceError,
} from "../ports/model-round.ts";
import {
  isModelRouteDurabilityError,
  ModelRouteRecoveredFailureError,
} from "../model-route/index.ts";
import {
  ModelProviderRequestError,
  diagnosticDetails,
  safeRuntimeFailure,
} from "../../../integrations/providers/provider-errors.ts";
import { createModelRouteRuntimeHooks } from "./model-route-runtime-hooks.ts";
import { isPhaseScopedMemoryProjectionError } from
  "../../context/context-projection.ts";
import { operationalFailureMessage } from "./turn-runtime-failure.ts";
import { guidedFinalTransition } from "./guided-final-transition.ts";
import {
  createNoopRuntimeMemoryAttributionPort,
  type RuntimeMemoryAttributionPort,
} from "../../../operations/diagnostics/runtime-memory-attribution/index.ts";
import {
  createNoopTurnDeveloperLogCapturePort,
  type TurnDeveloperLogCapturePort,
} from "../../../operations/diagnostics/developer-log-turn-capture/index.ts";

export type TurnRuntimeDependencies = {
  admission: TurnAdmissionRepository;
  turns: TurnStateRepository;
  messages: CanonicalMessageStore;
  agent: BtccAgentLoop;
  memoryAttribution?: RuntimeMemoryAttributionPort;
  developerLogCapture?: TurnDeveloperLogCapturePort;
  progress?: BtccTurnProgressObserver;
  committedSuccessorReadiness?: CommittedSuccessorReadiness;
};

/**
 * The only lifecycle runtime for a BTCC Turn.
 *
 * This module deliberately contains no model policy.  The model loop returns
 * a candidate final and this runtime owns durable admission, fencing,
 * canonical-message insertion, and delivery-state projection.
 */
class DefaultTurnRuntime implements BtccTurnRuntime {
  private readonly supervisor = createTurnExecutionSupervisor();
  private readonly activeTurns = new Map<string, Promise<BtccTurnOutcome>>();
  private readonly memoryAttribution: RuntimeMemoryAttributionPort;
  private readonly developerLogCapture: TurnDeveloperLogCapturePort;

  constructor(private readonly dependencies: TurnRuntimeDependencies) {
    this.memoryAttribution = dependencies.memoryAttribution ??
      createNoopRuntimeMemoryAttributionPort();
    this.developerLogCapture = dependencies.developerLogCapture ??
      createNoopTurnDeveloperLogCapturePort();
  }
  runTurn(
    command: BtccRunCommand,
    progress?: BtccTurnProgressObserver,
    onAdmitted?: (isFresh: boolean) => void | Promise<void>,
  ): Promise<BtccTurnOutcome> {
    const active = this.activeTurns.get(command.turnId);
    if (active) return active;
    const running = this.run(
      command,
      progress ?? this.dependencies.progress,
      onAdmitted,
    )
      .finally(() => this.activeTurns.delete(command.turnId));
    this.activeTurns.set(command.turnId, running);
    return running;
  }

  async stopTurn(command: BtccStopCommand): Promise<BtccTurnOutcome> {
    return stopTurn(command, this.dependencies.turns, this.supervisor);
  }

  private async run(
    command: BtccRunCommand,
    progress: BtccTurnProgressObserver | undefined,
    onAdmitted: ((isFresh: boolean) => void | Promise<void>) | undefined,
  ): Promise<BtccTurnOutcome> {
    this.memoryAttribution.checkpoint({ event: "turn_start", operation: "turn" });
    try {
      return await this.runTurnLifecycle(command, progress, onAdmitted);
    } finally {
      this.memoryAttribution.checkpoint({ event: "turn_end", operation: "turn" });
    }
  }

  private async runTurnLifecycle(
    command: BtccRunCommand,
    progress: BtccTurnProgressObserver | undefined,
    onAdmitted: ((isFresh: boolean) => void | Promise<void>) | undefined,
  ): Promise<BtccTurnOutcome> {
    let turn = await loadOrAdmitTurn(
      command,
      this.dependencies,
      (_admitted, isFresh) => onAdmitted?.(isFresh),
    );
    if (isTerminal(turn)) {
      await this.publishTerminal(progress, turn);
      return projectTerminalOutcome(turn);
    }
    if (turn.semanticState === "admitted") {
      await publishState(progress, turn);
    }
    if (turn.semanticState !== "delivery_committed") {
      turn = await this.runAgentAndCommit(turn, progress, command.recoveryAttempt);
    }
    if (turn.semanticState === "cancelled") {
      await this.publishTerminal(progress, turn);
      return projectTerminalOutcome(turn);
    }
    const delivered = await this.deliver(turn, progress);
    return projectTerminalOutcome(delivered);
  }

  private async runAgentAndCommit(
    turn: TurnRecord,
    progress: BtccTurnProgressObserver | undefined,
    recoveryAttempt?: number,
  ): Promise<TurnRecord> {
    const permit = this.supervisor.enter({
      turnId: turn.turnId,
      executionFence: turn.executionFence,
      semanticState: turn.semanticState,
    });
    try {
      const claim = await acquireStateExecution(
        turn,
        { ...this.dependencies, progress },
        permit,
      );
      let result: BtccAgentLoopResult;
      try {
        result = await this.dependencies.agent.run({
          turn,
          recoveryAttempt,
          signal: permit.signal,
          memoryAttribution: this.memoryAttribution,
          progress,
          ...createModelRouteRuntimeHooks({
            turn,
            claim,
            turns: this.dependencies.turns,
          }),
        });
        this.developerLogCapture.capture({
          kind: "model_turn",
          turn,
          result,
          timestamp: new Date().toISOString(),
        });
      } catch (error) {
        this.developerLogCapture.capture({
          kind: "model_turn_error",
          turn,
          failure: safeRuntimeFailure(error),
          diagnostics: diagnosticDetails(error),
          timestamp: new Date().toISOString(),
        });
        if (isModelRouteDurabilityError(error) ||
            isPhaseContinuityProjectionError(error) ||
            isPhaseScopedMemoryProjectionError(error) ||
            isRoundToolSurfaceError(error) ||
            isGuidedWorkCloseoutError(error)) throw error;
        if (isRetryableProviderExhaustion(error)) {
          const failure = safeRuntimeFailure(error);
          const failureCode = error instanceof ModelRouteRecoveredFailureError
            ? error.failureCode
            : failure.code;
          await progress?.runtimeFaulted?.({
            turnId: turn.turnId,
            sessionId: turn.sessionId,
            faultId: `${turn.turnId}:provider-transport-exhausted`,
            kind: "provider_transport_exhausted",
            retryable: true,
            publicSummary: operationalFailureMessage(turn.originalMessage, error),
            operatorSummary: `Provider recovery exhausted (${failureCode}).`,
            safeErrorCode: failureCode,
            createdAt: new Date().toISOString(),
          });
          throw error;
        }
        permit.assertActive();
        result = {
          route: "assisted",
          content: operationalFailureMessage(turn.originalMessage, error),
        };
      }
      permit.assertActive();
      const transition = guidedFinalTransition(turn, result);
      while (true) {
        try {
          await this.dependencies.turns.commitTransition({ turn, claim, transition });
          break;
        } catch (error) {
          if (!isSqliteContention(error) || !this.dependencies.committedSuccessorReadiness) {
            throw error;
          }
          await this.dependencies.committedSuccessorReadiness
            .waitForStorageReadiness(permit.signal);
          permit.assertActive();
        }
      }
      const committed = await this.dependencies.turns.activateCommittedSuccessor(turn.turnId);
      await this.publishDeliveryState(progress, committed);
      return committed;
    } catch (error) {
      const current = await this.dependencies.turns.findTurn(turn.turnId).catch(() => null);
      if (current?.semanticState === "cancelled") return current;
      throw error;
    } finally {
      permit.close();
    }
  }

  private async deliver(
    initial: TurnRecord,
    progress: BtccTurnProgressObserver | undefined,
  ): Promise<TurnRecord> {
    let turn = initial;
    while (turn.semanticState === "delivery_committed") {
      const permit = this.supervisor.enter({
        turnId: turn.turnId,
        executionFence: turn.executionFence,
        semanticState: turn.semanticState,
      });
      try {
        const claim = await acquireStateExecution(
          turn,
          { ...this.dependencies, progress },
          permit,
        );
        const message = await insertCanonicalMessage({
          turn,
          messages: this.dependencies.messages,
        });
        permit.assertActive();
        await this.dependencies.turns.commitTransition({
          turn,
          claim,
          transition: {
            kind: "observe_delivery",
            successor: "delivered",
            assistantMessageId: message.messageId,
          },
        });
        turn = await this.dependencies.turns.activateCommittedSuccessor(turn.turnId);
      } catch (error) {
        const current = await this.dependencies.turns.findTurn(turn.turnId).catch(() => null);
        if (permit.signal.aborted && current?.semanticState === "delivery_committed") {
          turn = current;
          continue;
        }
        if (current?.semanticState === "cancelled" || current?.semanticState === "delivered") {
          turn = current;
          break;
        }
        throw error;
      } finally {
        permit.close();
      }
    }
    if (!isTerminal(turn)) {
      throw new Error(`BTCC Turn reached delivery from ${turn.semanticState}`);
    }
    await this.publishTerminal(progress, turn);
    return turn;
  }

  private async publishDeliveryState(
    progress: BtccTurnProgressObserver | undefined,
    turn: TurnRecord,
  ): Promise<void> {
    if (turn.semanticState !== "delivery_committed") return;
    await publishState(progress, turn);
  }

  private async publishTerminal(
    progress: BtccTurnProgressObserver | undefined,
    turn: TurnRecord,
  ): Promise<void> {
    if (!isTerminal(turn)) return;
    this.memoryAttribution.terminal(
      turn.semanticState === "cancelled" ? "cancelled" : "delivered",
    );
    await publishState(progress, turn);
  }
}

function isRetryableProviderExhaustion(error: unknown): boolean {
  if (error instanceof ModelProviderRequestError) return error.retryable;
  return error instanceof ModelRouteRecoveredFailureError &&
    error.disposition === "retry";
}

export function createTurnRuntime(dependencies: TurnRuntimeDependencies): BtccTurnRuntime {
  return new DefaultTurnRuntime(dependencies);
}

async function publishState(
  observer: BtccTurnProgressObserver | undefined,
  turn: TurnRecord,
): Promise<void> {
  if (!observer) return;
  try {
    await observer.stateChanged({
      turnId: turn.turnId,
      semanticState: turn.semanticState,
      turnRevision: turn.revision,
    });
  } catch {
    // Public progress cannot veto durable Turn state.
  }
}

function isTerminal(turn: TurnRecord): boolean {
  return turn.semanticState === "delivered" || turn.semanticState === "cancelled";
}

File: packages/butler-agent/rust/agent/src/btcc/turn/runtime.rs
use std::sync::Arc;

use super::contracts::{
    ProgressEvent, StopPersistenceOutcome, SuspensionReason, TurnRecord, TurnSemanticState,
    TurnTransition,
};
use super::failure::runtime_failure_message;
use super::ports::{
    AgentLoop, AgentLoopError, AgentLoopProgress, CanonicalMessageStore, PreparedConversation,
    ProgressEventRepository, StorageReadiness, TransitionCommitError, TurnStore,
};
use super::progress::TurnProgressScope;
use super::supervisor::TurnExecutionSupervisor;
use super::transition::guided_final;
use crate::btcc::BtccCode;
use crate::btcc::{BtccError, DeliveredOutcome, ProgressDestination, TurnOutcome, TurnOutcomeKind};

pub(super) struct TurnRuntime {
    store: Arc<dyn TurnStore>,
    agent: Arc<dyn AgentLoop>,
    messages: Arc<dyn CanonicalMessageStore>,
    progress: Arc<dyn ProgressEventRepository>,
    readiness: Arc<dyn StorageReadiness>,
    developer_log_capture: Arc<dyn super::ports::TurnDeveloperLogCapturePort>,
    supervisor: TurnExecutionSupervisor,
}

impl TurnRuntime {
    pub(super) fn new(
        store: Arc<dyn TurnStore>,
        agent: Arc<dyn AgentLoop>,
        messages: Arc<dyn CanonicalMessageStore>,
        progress: Arc<dyn ProgressEventRepository>,
        readiness: Arc<dyn StorageReadiness>,
        developer_log_capture: Arc<dyn super::ports::TurnDeveloperLogCapturePort>,
        supervisor: TurnExecutionSupervisor,
    ) -> Self {
        Self {
            store,
            agent,
            messages,
            progress,
            readiness,
            developer_log_capture,
            supervisor,
        }
    }

    pub(super) async fn run_prepared(
        &self,
        prepared: &super::contracts::PreparedTurn,
        conversation: &dyn PreparedConversation,
        destination: ProgressDestination,
    ) -> Result<(TurnOutcome, bool), BtccError> {
        let (mut turn, fresh) = self.store.load_or_admit(prepared).await?;
        if fresh && prepared.is_fresh && turn.semantic_state != TurnSemanticState::Cancelled {
            self.publish(conversation, &turn, &destination, ProgressEvent::Started)
                .await?;
        }
        if terminal(turn.semantic_state) {
            self.publish_state(conversation, &turn, &destination).await;
            return Ok((project_terminal(&turn)?, fresh));
        }
        if turn.suspension == Some(SuspensionReason::AuthorityPending) {
            turn = self
                .store
                .resume_authority(&turn.turn_id)
                .await?
                .unwrap_or(turn);
        }
        if let Some(reason) = turn.suspension {
            return Ok((suspended(&turn.turn_id, reason), fresh));
        }
        if turn.semantic_state == TurnSemanticState::Admitted {
            self.publish_state(conversation, &turn, &destination).await;
        }
        if turn.semantic_state != TurnSemanticState::DeliveryCommitted {
            let recovery_turn_id = turn.turn_id.clone();
            turn = match self
                .run_agent(
                    turn,
                    conversation,
                    &destination,
                    prepared.request.recovery_attempt.unwrap_or(1),
                )
                .await
            {
                Ok(turn) => turn,
                Err(error) => match self.store.find_turn(&recovery_turn_id).await {
                    Ok(Some(current)) if current.semantic_state == TurnSemanticState::Cancelled => {
                        current
                    }
                    _ => return Err(error),
                },
            };
        }
        if let Some(reason) = turn.suspension {
            return Ok((suspended(&turn.turn_id, reason), fresh));
        }
        if turn.semantic_state == TurnSemanticState::Cancelled {
            self.supervisor.observe_terminal(&turn.turn_id);
            self.publish_state(conversation, &turn, &destination).await;
            return Ok((project_terminal(&turn)?, fresh));
        }
        let delivered = self.deliver(turn, conversation, &destination).await?;
        Ok((project_terminal(&delivered)?, fresh))
    }

    pub(super) async fn stop(&self, turn_id: &str) -> Result<TurnOutcome, BtccError> {
        // The synchronous fence precedes the first repository await.
        let ticket = self.supervisor.install_stop(turn_id);
        match self.store.stop(turn_id).await {
            Ok(outcome) => {
                self.supervisor.observe_stop(&ticket, &outcome.clone());
                Ok(TurnOutcome {
                    result: stop_outcome(turn_id, outcome),
                    admission: None,
                })
            }
            Err(_) => {
                self.supervisor.observe_stop_failure(&ticket);
                Ok(TurnOutcome {
                    result: TurnOutcomeKind::FencedPendingPersistence {
                        turn_id: turn_id.to_owned(),
                    },
                    admission: None,
                })
            }
        }
    }

    async fn run_agent(
        &self,
        turn: TurnRecord,
        conversation: &dyn PreparedConversation,
        destination: &ProgressDestination,
        recovery_attempt: u32,
    ) -> Result<TurnRecord, BtccError> {
        let permit = self.supervisor.enter(&turn.turn_id, turn.semantic_state)?;
        let claim = self.store.acquire_state_claim(&turn).await?;
        let progress = self.progress_scope(conversation, &turn, destination);
        let developer_log = self.developer_log_capture.start_execution();
        let agent_result = self
            .agent
            .run(
                &turn,
                &claim,
                recovery_attempt,
                &progress,
                developer_log.as_ref(),
                permit.cancellation(),
            )
            .await;
        developer_log.capture(&turn, agent_result.as_ref()).await;
        let mut result = match agent_result {
            Ok(result) => result,
            Err(AgentLoopError::Propagate(error)) => return Err(error),
            Err(AgentLoopError::Runtime(failure)) => super::contracts::AgentLoopResult {
                route: super::contracts::ExecutionRoute::Assisted,
                content: runtime_failure_message(&turn.original_message, &failure),
                terminal_outcome: None,
                suspension: None,
                authority_continuation: None,
                work_status: None,
                accepted_work_result: Some(crate::btcc::AcceptedWorkResult {
                    status: crate::btcc::AcceptedWorkStatus::Failed,
                }),
                runtime_failure: Some(failure),
                artifacts: Vec::new(),
                changed_files: Vec::new(),
                plan: None,
                model_identity: None,
            },
        };
        permit.assert_active()?;
        if let Some(failure) = &result.runtime_failure {
            let work_completed = result
                .accepted_work_result
                .as_ref()
                .is_some_and(|value| value.status == crate::btcc::AcceptedWorkStatus::Success);
            result.content = super::failure::runtime_failure_message_for_work(
                &turn.original_message,
                failure,
                work_completed,
            );
            result
                .accepted_work_result
                .get_or_insert(crate::btcc::AcceptedWorkResult {
                    status: crate::btcc::AcceptedWorkStatus::Failed,
                });
        }
        let transition = if let Some(reason) = result.suspension {
            TurnTransition::Suspend {
                reason,
                authority_continuation: result.authority_continuation,
            }
        } else {
            guided_final(&turn, result)?
        };
        if matches!(transition, TurnTransition::Suspend { .. }) {
            self.store
                .commit_transition(&turn, &claim, &transition)
                .await
                .map_err(commit_error)?;
        } else {
            loop {
                match self
                    .store
                    .commit_transition(&turn, &claim, &transition)
                    .await
                {
                    Ok(()) => break,
                    Err(TransitionCommitError::Contention) => {
                        self.readiness.wait(permit.cancellation()).await?;
                        permit.assert_active()?;
                    }
                    Err(TransitionCommitError::Failure(error)) => return Err(error),
                }
            }
        }
        let committed = self.store.activate_successor(&turn.turn_id).await?;
        if let TurnTransition::Suspend { reason, .. } = transition {
            if committed.suspension != Some(reason) {
                return Err(BtccError::detected(
                    BtccCode::SuspensionNotPersisted,
                    "BTCC suspension commit did not persist its reason",
                ));
            }
            return Ok(committed);
        }
        self.publish_state(conversation, &committed, destination)
            .await;
        Ok(committed)
    }

    async fn deliver(
        &self,
        mut turn: TurnRecord,
        conversation: &dyn PreparedConversation,
        destination: &ProgressDestination,
    ) -> Result<TurnRecord, BtccError> {
        while turn.semantic_state == TurnSemanticState::DeliveryCommitted {
            let permit = self.supervisor.enter(&turn.turn_id, turn.semantic_state)?;
            let attempt = async {
                let claim = self.store.acquire_state_claim(&turn).await?;
                let message_id = self.messages.insert(&turn).await?;
                permit.assert_active()?;
                self.store
                    .commit_transition(
                        &turn,
                        &claim,
                        &TurnTransition::ObserveDelivery {
                            assistant_message_id: message_id,
                        },
                    )
                    .await
                    .map_err(commit_error)?;
                self.store.activate_successor(&turn.turn_id).await
            }
            .await;
            match attempt {
                Ok(successor) => turn = successor,
                Err(error) => {
                    let current = self.store.find_turn(&turn.turn_id).await.unwrap_or(None);
                    match current {
                        Some(current)
                            if permit.cancellation().is_cancelled()
                                && current.semantic_state
                                    == TurnSemanticState::DeliveryCommitted =>
                        {
                            turn = current;
                        }
                        Some(current) if terminal(current.semantic_state) => {
                            turn = current;
                            break;
                        }
                        _ => return Err(error),
                    }
                }
            }
        }
        if !terminal(turn.semantic_state) {
            return Err(BtccError::detected(
                BtccCode::InvalidDeliveryState,
                "Turn did not reach a terminal delivery state",
            ));
        }
        self.supervisor.observe_terminal(&turn.turn_id);
        self.publish_state(conversation, &turn, destination).await;
        Ok(turn)
    }

    async fn publish_state(
        &self,
        conversation: &dyn PreparedConversation,
        turn: &TurnRecord,
        destination: &ProgressDestination,
    ) {
        let _ = self
            .publish(
                conversation,
                turn,
                destination,
                ProgressEvent::StateChanged {
                    semantic_state: turn.semantic_state,
                    turn_revision: turn.revision,
                },
            )
            .await;
    }

    async fn publish(
        &self,
        conversation: &dyn PreparedConversation,
        turn: &TurnRecord,
        destination: &ProgressDestination,
        event: ProgressEvent,
    ) -> Result<(), BtccError> {
        self.progress_scope(conversation, turn, destination)
            .emit(crate::btcc::progress::project(event))
            .await
    }

    fn progress_scope<'a>(
        &'a self,
        conversation: &'a dyn PreparedConversation,
        turn: &'a TurnRecord,
        destination: &'a ProgressDestination,
    ) -> TurnProgressScope<'a> {
        TurnProgressScope {
            conversation,
            repository: self.progress.as_ref(),
            session_id: &turn.session_id,
            turn_id: &turn.turn_id,
            destination,
        }
    }
}

fn project_terminal(turn: &TurnRecord) -> Result<TurnOutcome, BtccError> {
    let result = if turn.semantic_state == TurnSemanticState::Cancelled {
        TurnOutcomeKind::Cancelled {
            turn_id: turn.turn_id.clone(),
        }
    } else {
        let payload = turn.final_payload.as_ref().ok_or_else(|| {
            BtccError::detected(
                BtccCode::MissingFinalPayload,
                "Delivered Turn has no final payload",
            )
        })?;
        let message_id = turn.canonical_assistant_message_id.clone().ok_or_else(|| {
            BtccError::detected(
                BtccCode::MissingCanonicalMessage,
                "Delivered Turn has no message",
            )
        })?;
        TurnOutcomeKind::Delivered(Box::new(DeliveredOutcome {
            turn_id: turn.turn_id.clone(),
            message_id,
            content: payload.content.clone(),
            work_status: payload.work_status,
            accepted_work_result: payload.accepted_work_result.clone(),
            runtime_failure: payload.runtime_failure.clone(),
            execution_outcome: payload.execution_outcome,
            artifacts: payload.artifacts.clone(),
            changed_files: payload.changed_files.clone(),
            plan: payload.plan.clone(),
            model_identity: payload.model_identity.clone(),
        }))
    };
    Ok(TurnOutcome {
        result,
        admission: None,
    })
}

fn stop_outcome(turn_id: &str, outcome: StopPersistenceOutcome) -> TurnOutcomeKind {
    match outcome {
        StopPersistenceOutcome::Cancelled => TurnOutcomeKind::Cancelled {
            turn_id: turn_id.into(),
        },
        StopPersistenceOutcome::AlreadyCancelled => TurnOutcomeKind::AlreadyCancelled {
            turn_id: turn_id.into(),
        },
        StopPersistenceOutcome::AlreadyFinalizing => TurnOutcomeKind::AlreadyFinalizing {
            turn_id: turn_id.into(),
        },
        StopPersistenceOutcome::AlreadyDelivered(delivered) => {
            TurnOutcomeKind::AlreadyDelivered(delivered)
        }
    }
}

fn suspended(turn_id: &str, reason: SuspensionReason) -> TurnOutcome {
    TurnOutcome {
        result: TurnOutcomeKind::Suspended {
            turn_id: turn_id.into(),
            reason: match reason {
                SuspensionReason::AuthorityPending => "authority_pending",
                SuspensionReason::WaitingForWorker => "waiting_for_worker",
            }
            .into(),
        },
        admission: None,
    }
}

fn terminal(state: TurnSemanticState) -> bool {
    matches!(
        state,
        TurnSemanticState::Delivered | TurnSemanticState::Cancelled
    )
}

fn commit_error(error: TransitionCommitError) -> BtccError {
    match error {
        TransitionCommitError::Contention => {
            BtccError::detected(BtccCode::SqliteContention, "delivery transition contention")
        }
        TransitionCommitError::Failure(error) => error,
    }
}
